File: src/main/java/com/pyamsoft/padlock/dagger/wrapper/JobSchedulerCompatImpl.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.padlock.dagger.wrapper;

import android.app.Service;
import android.content.Context;
import android.os.Build;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import com.birbit.android.jobqueue.Job;
import com.birbit.android.jobqueue.JobManager;
import com.birbit.android.jobqueue.TagConstraint;
import com.birbit.android.jobqueue.config.Configuration;
import com.birbit.android.jobqueue.log.CustomLogger;
import com.birbit.android.jobqueue.scheduling.FrameworkJobSchedulerService;
import com.birbit.android.jobqueue.scheduling.GcmJobSchedulerService;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.pyamsoft.padlock.BuildConfig;
import com.pyamsoft.padlock.app.service.job.PadLockFrameworkJobSchedulerService;
import com.pyamsoft.padlock.app.service.job.PadLockGCMJobSchedulerService;
import javax.inject.Inject;
import timber.log.Timber;

class JobSchedulerCompatImpl implements JobSchedulerCompat {

  @NonNull private final JobManager jobManager;

  @Inject JobSchedulerCompatImpl(@NonNull Context context) {
    jobManager = createJobManager(context.getApplicationContext());
  }

  @VisibleForTesting @CheckResult @NonNull JobManager createJobManager(@NonNull Context context) {
    final Configuration.Builder builder =
        new Configuration.Builder(context.getApplicationContext()).customLogger(new CustomLogger() {
          @Override public boolean isDebugEnabled() {
            return BuildConfig.DEBUG;
          }

          @Override public void d(String text, Object... args) {
            Timber.d(text, args);
          }

          @Override public void e(Throwable t, String text, Object... args) {
            Timber.e(t, text, args);
          }

          @Override public void e(String text, Object... args) {
            Timber.e(text, args);
          }

          @Override public void v(String text, Object... args) {
            Timber.v(text, args);
          }
        });

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
      Timber.d("Create scheduler using JobScheduler framework");
      builder.scheduler(FrameworkJobSchedulerService.createSchedulerFor(context,
          PadLockFrameworkJobSchedulerService.class));
    } else {
      final int googleAvailable =
          GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context);
      if (googleAvailable == ConnectionResult.SUCCESS) {
        Timber.d("Create scheduler using Google play services");

        // Batch by default
        builder.scheduler(GcmJobSchedulerService.createSchedulerFor(context,
            PadLockGCMJobSchedulerService.class));
      } else {
        Timber.e("Could not create a scheduler to use with the JobScheduler");
      }
    }

    Timber.d("Create a new JobManager");
    return new JobManager(builder.build());
  }

  @Override @NonNull public JobManager provideManagerToService(@NonNull Service service) {
    Timber.d("Provide JobManager instance to service: %s", service);
    return jobManager;
  }

  @Override
  public void cancelJobsInBackground(@NonNull TagConstraint constraint, @NonNull String... tags) {
    jobManager.cancelJobsInBackground(null, constraint, tags);
  }

  @Override public void cancelJobs(@NonNull TagConstraint constraint, @NonNull String... tags) {
    jobManager.cancelJobs(constraint, tags);
  }

  @Override public void addJob(@NonNull Job job) {
    jobManager.addJob(job);
  }
}

File: padlock-base/src/main/java/com/pyamsoft/padlock/base/jobs/JobSchedulerCompatImpl.kt
/*
 * Copyright 2019 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.padlock.base.jobs

import android.app.job.JobInfo
import android.app.job.JobScheduler
import android.app.job.JobService
import android.content.ComponentName
import android.content.Context
import android.os.PersistableBundle
import androidx.annotation.CheckResult
import com.pyamsoft.padlock.api.service.JobSchedulerCompat
import com.pyamsoft.padlock.api.service.JobSchedulerCompat.JobType
import com.pyamsoft.padlock.api.service.JobSchedulerCompat.JobType.RECHECK
import com.pyamsoft.padlock.api.service.JobSchedulerCompat.JobType.SERVICE_TEMP_PAUSE
import com.pyamsoft.padlock.model.service.Recheck
import timber.log.Timber
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
internal class JobSchedulerCompatImpl @Inject internal constructor(
  private val context: Context,
  private val jobServiceClass: Class<out JobService>
) : JobSchedulerCompat {

  private val serviceComponent by lazy { ComponentName(context, jobServiceClass) }
  private val recheckCache by lazy { ConcurrentHashMap<String, Int>() }
  private val jobScheduler by lazy {
    context.getSystemService(Context.JOB_SCHEDULER_SERVICE) as JobScheduler
  }

  @CheckResult
  private fun generateRecheckKey(
    extras: Map<String, String>
  ): String {
    val packageName = requireNotNull(extras[Recheck.EXTRA_PACKAGE_NAME])
    val className = requireNotNull(extras[Recheck.EXTRA_CLASS_NAME])
    return "$packageName|$className"
  }

  @CheckResult
  private fun buildJobInfo(
    jobId: Int,
    type: JobType,
    triggerAfter: Long,
    extras: Map<String, String>
  ): JobInfo {
    val builder = JobInfo.Builder(jobId, serviceComponent)
        // Job will run anywhere from CURRENT + triggerTime
        .setMinimumLatency(triggerAfter)
        // Job will run by CURRENT + (2 * triggerTime)
        .setOverrideDeadline(2 * triggerAfter)
        // Do not persist across restart
        .setPersisted(false)
        // Bundle extras
        .setExtras(PersistableBundle().apply {
          putString(JobSchedulerCompat.KEY_JOB_TYPE, type.name)
          for ((key, value) in extras) {
            putString(key, value)
          }
        })

    return builder.build()
  }

  override fun cancelAll() {
    // Cancel all of the Recheck jobs but do not cancel the TEMP SERVICE job
    // since it will need to restart the service at a later point
    recheckCache.values.forEach { jobScheduler.cancel(it) }
    recheckCache.clear()
  }

  override fun cancel(
    type: JobType,
    extras: Map<String, String>
  ) {
    when (type) {
      // Service temp pause must have an active id to cancel
      SERVICE_TEMP_PAUSE -> {
        Timber.d("Cancel job for temp pause service")
        jobScheduler.cancel(SERVICE_PAUSE_JOB_ID)
      }
      RECHECK -> {
        // Check for key in cache map and cancel job
        val key = generateRecheckKey(extras)
        Timber.d("Cancel job for recheck: $key")
        recheckCache[key]?.also { jobScheduler.cancel(it) }
      }
    }
  }

  override fun queue(
    type: JobType,
    triggerAfter: Long,
    extras: Map<String, String>
  ) {
    // Cancel old jobs
    cancel(type, extras)

    val job: JobInfo
    when (type) {
      SERVICE_TEMP_PAUSE -> {
        Timber.d("Construct job for temp pause service")
        job = buildJobInfo(SERVICE_PAUSE_JOB_ID, type, triggerAfter, extras)
      }
      RECHECK -> {
        val key = generateRecheckKey(extras)
        Timber.d("Construct job for recheck: $key")
        job = buildJobInfo(key.hashCode(), type, triggerAfter, extras)
      }
    }

    jobScheduler.schedule(job)
  }

  companion object {

    private const val SERVICE_PAUSE_JOB_ID = 1

  }
}
